File: 02.Queue/01.2.TreeLevelOrder.py
# 本题的测试平台链接：
# https://leetcode-cn.com/problems/binary-tree-level-order-traversal/

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        # 存放每一层遍历的结果
        ans = []

        # 记录当前层，注意不要把空的root放进去
        cur = []
        if root:
            cur.append(root)
        
        while len(cur):
            # nextLevel用来存放下一层的结点
            nextLevel = []

            # curResult用来存放当前层的值
            curResult = []

            for c in cur:
                # 把当前层的值放到curResult里面
                curResult.append(c.val)

                # 生成下一层
                if c.left:
                    nextLevel.append(c.left)
                if c.right:
                    nextLevel.append(c.right)
            
            # 指向下一层
            cur = nextLevel

            # 把当前层放到结果中
            ans.append([x for x in curResult])
        
        # 返回结果
        return ans

File: 02.Queue/01.2.TreeLevelOrder.java
// 测试平台链接：https://leetcode-cn.com/problems/binary-tree-level-order-traversal/submissions/
// LeetCode 102
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();

        // 初始化当前层结点
        List<TreeNode> curLevel = new ArrayList<>();
        // 注意：需要root不空的时候才加到里面。
        if (root != null) {
            curLevel.add(root);
        }

        while (curLevel.size() > 0) {
            // 准备用来存放下一层的结点
            List<TreeNode> nextLevel = new ArrayList<>();
            // 用来存放当前层的结果
            List<Integer> curResult = new ArrayList<>();

            // 遍历当前层的每个结点
            for (TreeNode cur: curLevel) {
                // 把当前层的值存放到当前结果里面
                curResult.add(cur.val);

                // 生成下一层
                if (cur.left != null) {
                    nextLevel.add(cur.left);
                }
                if (cur.right != null) {
                    nextLevel.add(cur.right);
                }
            }

            // 注意这里的更迭!
            curLevel = nextLevel;

            ans.add(curResult);
        }

        return ans;
    }
}
